### PR TITLE
45 implement legal moves updated

### LINKED ISSUE #45: Implement legal moves
Legalize moves

### PR DESCRIPTION
**Pull request by abrahimt**
Made the basic legal moves function.

### CODE REVIEW

File: src/board.rs [hunk 1]
```diff
@@ -486,6 +486,138 @@ impl Chessboard {
...
+        {
+            return false;
+        }
+        // Extract old and new square coordinates
+        if let (Some(old_file), Some(old_rank), Some(new_file), Some(new_rank)) = (
+            current_pos.chars().next(),
+            current_pos.chars().next_back(),
+            new_pos.chars().next(),
+            new_pos.chars().next_back(),
+        ) {
```
Comment (chanson02) [inline src/board.rs]: invert this and return early

File: src/board.rs [hunk 2]
```diff
@@ -486,6 +486,138 @@ impl Chessboard {
...
+                'q' => is_legal = Chessboard::legal_queen(old_square, new_square),
+                'n' => is_legal = Chessboard::legal_knight(old_square, new_square),
+                'P' => is_legal = Chessboard::legal_pawn(cb, old_square, new_square),
+                'R' => is_legal = Chessboard::legal_rook(old_square, new_square),
+                'B' => is_legal = Chessboard::legal_bishop(old_square, new_square),
+                'K' => is_legal = Chessboard::legal_king(old_square, new_square),
+                'Q' => is_legal = Chessboard::legal_queen(old_square, new_square),
+                'N' => is_legal = Chessboard::legal_knight(old_square, new_square),
+                _ => return false,
+            }
```
Comment (chanson02) [inline src/board.rs]: return the result of this match, instead of updating `is_legal`

File: src/board.rs [hunk 3]
```diff
@@ -486,6 +486,138 @@ impl Chessboard {
...
+                        new_rank.to_digit(10).unwrap() as usize,
+                        (new_file as u8 - b'A') as usize,
+                    )
+                    .is_lowercase())
+                || (piece.is_uppercase()
+                    & cb.piece_at_position(
+                        new_rank.to_digit(10).unwrap() as usize,
+                        (new_file as u8 - b'A') as usize,
+                    )
+                    .is_uppercase())
```
Comment (chanson02) [inline src/board.rs]: this is a lot of unprotected unwraps, are we certain this won't fail?

File: src/board.rs [hunk 4]
```diff
@@ -486,6 +486,138 @@ impl Chessboard {
...
+            || new_square == (old_square - 15)
+    }
+    fn legal_bishop(old_square: u64, new_square: u64) -> bool {
+        //add and subtract multiples of 7 and 9 (max 8 long diagonal)
+        if new_square > old_square {
+            (new_square - old_square) % 7 == 0 || (new_square - old_square) % 9 == 0
+        } else {
+            (old_square - new_square) % 7 == 0 || (old_square - new_square) % 9 == 0
+        }
+    }
```
Comment (chanson02) [inline src/board.rs]: smart.

File: src/board.rs [hunk 5]
```diff
@@ -486,6 +486,138 @@ impl Chessboard {
...
+    fn legal_rook(old_square: u64, new_square: u64) -> bool {
+        if new_square % 8 == old_square % 8 || new_square / 8 == old_square / 8 {
+            return true;
+        }
+        false
+    }
```
Comment (chanson02) [inline src/board.rs]: also smart. however we can remove the conditional and just return
`new_square % 8 == old_square % 8 || new_square / 8 == old_square / 8`

File: src/board.rs [hunk 6]
```diff
@@ -486,6 +486,138 @@ impl Chessboard {
...
+            || new_square == old_square + 8 //up
+            || new_square == old_square - 8 //down
+            || new_square == old_square + 9 //diag up right
+            || new_square == old_square - 9 //diag down left
+            || new_square == old_square - 7 //diag down right
+    }
+    fn legal_queen(old_square: u64, new_square: u64) -> bool {
+        Chessboard::legal_bishop(old_square, new_square)
+            && Chessboard::legal_rook(old_square, new_square)
+    }
```
Comment (chanson02) [inline src/board.rs]: genius.

**Review by chanson02 — CHANGES_REQUESTED** [review]

File: src/board.rs (hunk 4 above)
Comment (abrahimt) [reply src/board.rs]: You can trick this function.

File: src/board.rs [hunk 7]
```diff
@@ -271,74 +282,70 @@ impl Chessboard {
...
+                self.black_pawns |= add_new; // Set new position
+            }
+            'r' => {
+                self.black_rooks &= clear_old;
+                self.black_rooks |= add_new;
+            }
+            'b' => {
+                self.black_bishops &= clear_old;
+                self.black_bishops |= add_new;
+            }
```
Comment (chanson02) [inline src/board.rs:311]: nit:
```rust
let x = match piece {
  'p' => self.blac_pawns,
  'r' => self.black_rooks,
  etc
}
x &= clear_old; //clear old position
x |= add_new; //set new position
```

File: src/board.rs [hunk 8]
```diff
@@ -486,6 +493,42 @@ impl Chessboard {
...
+        Here is my proposition. We have two functions
+
+         is_valid_move_for_piece(piece, old_square, new_square)
+         just looks at the logic for how this piece moves (pawns can only move forward, etc)
+
+         is_legal_move(board, piece, old_sq, new_sq)
+         handles special cases that are specific to the game state (en passant, eating, etc.)
+         - is there already a piece at that position?
+         - is another piece blocking the path?
+    */
```
Comment (chanson02) [inline src/board.rs]: These comments shouldn't make it into main.

File: src/board/piece.rs [hunk 9]
```diff
@@ -0,0 +1,57 @@
+use crate::board::Chessboard;
+
+// I would love to cutdown the nesting here --Cooper
```
Comment (chanson02) [inline src/board/piece.rs]: This comment shouldn't make it onto main.

File: src/board/piece.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+pub fn legal_bishop(from: u64, to: u64) -> bool {
+    if to > from && (from % 2 == 0 && to % 2 == 0 || from % 2 != 0 && to % 2 != 0) {
+        (to - from) % 7 == 0 || (to - from) % 9 == 0
+    } else if to < from && (from % 2 == 0 && to % 2 == 0 || from % 2 != 0 && to % 2 != 0) {
+        (from - to) % 7 == 0 || (from - to) % 9 == 0
+    } else {
+        false
+    }
+}
```
Comment (chanson02) [inline src/board/piece.rs]: We should use some variable names here to increase readability if possible.

File: src/board/piece.rs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+pub fn legal_king(from: u64, to: u64) -> bool {
+    to == from + 1 //right
+    || to == from - 1 //left
+    || to == from + 8 //up
+    || to == from - 8 //down
+    || to == from + 9 //diag up right
+    || to == from - 9 //diag down left
+    || to == from - 7 //diag down right
+}
```
Comment (chanson02) [inline src/board/piece.rs]: This is going to crash if from is < the subtracting number.

File: src/board/piece.rs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+pub fn legal_knight(from: u64, to: u64) -> bool {
+    to == (from + 17)
+        || to == (from + 15)
+        || to == (from + 10)
+        || to == (from - 10)
+        || to == (from - 6)
+        || to == (from + 6)
+        || to == (from - 17)
+        || to == (from - 15)
+}
```
Comment (chanson02) [inline src/board/piece.rs]: protect against crashing.

**Review by chanson02 — CHANGES_REQUESTED** [review]

File: src/board.rs (hunk 7 above)
Comment (chanson02) [reply src/board.rs:311]: This won't work because the match goes into it's own scope.

File: src/board/piece.rs (hunk 10 above)
Comment (chanson02) [reply src/board/piece.rs]:
```rust
white_from = from % 2 == 0
white_to = to % 2 == 0
same_color = white_from == white_to

if same_color
  diff = (to - from).abs
  diff % 7 == 0 || diff % 9 == 0
else
  false
```

File: src/board/piece.rs (hunk 11 above)
Comment (chanson02) [reply src/board/piece.rs]: Actually this type function can be heavily simplified as well.
```rust
rank_diff = (from / 8 - to / 8).abs
file_diff = (from % 8 - to % 8).abs
rank_diff <= 1 && file_diff <= 1 && to != from
```

Comment (chanson02) [reply src/board/piece.rs]: You could potentially make that even easier by using `square_to_rank` and `square_to_file`!

File: src/board/piece.rs (hunk 12 above)
Comment (chanson02) [reply src/board/piece.rs]:
```rust
spaces_moved = (to - from).abs
match spaces {
  6 | 10 | 15 | 17 => true,
  _ => false
}
```

File: src/board/piece.rs [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    || to == from + 8 //up
+    || to == from - 8 //down
+    || to == from + 9 //diag up right
+    || to == from - 9 //diag down left
+    || to == from - 7 //diag down right
+}
+
+pub fn legal_queen(from: u64, to: u64) -> bool {
+    legal_bishop(from, to) && legal_rook(from, to)
+}
```
Comment (chanson02) [inline src/board/piece.rs:45]: I didn't make a test for this because I felt it wasn't necessary.

File: src/board/piece.rs [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    let direction = if white { 1 } else { -1 };
+    let initial_rank = if white { 2 } else { 7 };
+    let diff = to as i64 - from as i64;
+
+    if diff == 8 * direction {
+        true
+    } else {
+        rank == initial_rank && diff == 16 * direction
+    }
+}
```
Comment (chanson02) [inline src/board/piece.rs:14]: Low-key proud of this <3

File: src/board/piece.rs [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+        assert!(legal_knight(0, 17));
+        assert!(legal_knight(0, 15));
+        assert!(legal_knight(0, 10));
+        assert!(legal_knight(0, 6));
+
+        assert!(!legal_knight(0, 16));
+        assert!(!legal_knight(0, 8));
+        assert!(!legal_knight(0, 9));
+        assert!(!legal_knight(0, 7));
+    }
```
Comment (chanson02) [inline src/board/piece.rs]: This doesn't feel like a great way of testing this, but I can't think of anything better.

**Review by chanson02 — COMMENTED** [review]
Tests are implemented! The problem is only `legal_pawn` passes. Whoever wrote that function must be really smart!!!!!!

File: src/board/piece.rs (hunk 11 above)
Comment (abrahimt) [reply src/board/piece.rs]: You can't use .abs() on u64. Let's make an issue for this to see if we should change our squares to be signed.

File: src/board/piece.rs (hunk 12 above)
Comment (abrahimt) [reply src/board/piece.rs]: Same thing here about the signed numbers

File: src/board/piece.rs (hunk 15 above)
Comment (abrahimt) [reply src/board/piece.rs]: We should be testing through the is_valid_move_for_piece function.

File: src/board.rs [hunk 16]
```diff
@@ -486,6 +493,30 @@ impl Chessboard {
...
+            'B' => piece::legal_bishop(cur_square, new_square),
+            'k' => piece::legal_king(cur_square, new_square),
+            'K' => piece::legal_king(cur_square, new_square),
+            'q' => piece::legal_queen(cur_square, new_square),
+            'Q' => piece::legal_queen(cur_square, new_square),
+            'n' => piece::legal_knight(cur_square, new_square),
+            'N' => piece::legal_knight(cur_square, new_square),
+            _ => false,
+        }
+    }
```
Comment (chanson02) [inline src/board.rs:530]: Docstring

**Review by chanson02 — CHANGES_REQUESTED** [review]

**Review by chanson02 — APPROVED** [review]

**Pull request merged**